### PR TITLE
feat: Add proper example of configuring Rust functions.

### PR DESCRIPTION
**Pull request by dejanb**
# Changes

-:gift: As discussed in #430, here's the concrete example on how to use function configuration.

/kind feature

It uses a simple configuration struct to provide users with easy to adapt example on how to provide configuration to their functions.

### CODE REVIEW

**Review by lance — COMMENTED** [review]
Looks good to me. Though again, I'm not a rustafarian. Would love for @jcrossley3 to have a look.

File: templates/rust/events/src/handler.rs [hunk 1]
```diff
@@ -1,16 +1,21 @@
...
+    config: web::Data<HandlerConfig>,
+) -> Result<Event, actix_web::Error> {
     info!("event: {}", event);
 
     let input = match event.data() {
         Some(Data::Binary(v)) => from_slice(v)?,
         Some(Data::String(v)) => from_str(v)?,
         Some(Data::Json(v)) => v.to_owned(),
-        None => json!({ "name": "world" }),
+        None => json!({ "name": config.name }),
```
Comment (jcrossley3) [inline templates/rust/events/src/handler.rs:18]: Very nice!

File: templates/rust/events/src/handler.rs [hunk 2]
```diff
@@ -29,7 +34,7 @@ mod tests {
     async fn valid_input() {
         let mut input = Event::default();
         input.set_data("application/json", json!({"name": "bootsy"}));
-        let resp = handle(input).await;
+        let resp = handle(input, web::Data::new(HandlerConfig::default())).await;
```
Comment (jcrossley3) [inline templates/rust/events/src/handler.rs:37]: I really wish rust supported optional arguments. :)

Can we put this in a static constant local to the test module and call it `config` or something? Just so it won't look so gross?

File: templates/rust/events/src/handler.rs [hunk 3]
```diff
@@ -39,7 +44,7 @@ mod tests {
 
     #[actix_rt::test]
     async fn no_input() {
-        let resp = handle(Event::default()).await;
+        let resp = handle(Event::default(), web::Data::new(HandlerConfig::default())).await;
```
Comment (jcrossley3) [inline templates/rust/events/src/handler.rs:47]: ibid

File: templates/rust/http/src/handler.rs [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
 // Implement your function's logic here
-pub async fn index(req: HttpRequest) -> HttpResponse {
+pub async fn index(req: HttpRequest, config: web::Data<HandlerConfig>) -> HttpResponse {
     info!("{:#?}", req);
     if req.method() == Method::GET {
-        HttpResponse::Ok().body("Hello!\n")
+        HttpResponse::Ok().body(format!("Hello {}!\n", config.name))
     } else {
-        HttpResponse::Ok().body("Thanks!\n")
+        HttpResponse::Ok().body(format!("Thanks {}!\n", config.name))
```
Comment (jcrossley3) [inline templates/rust/http/src/handler.rs:11]: This kinda makes me want to add a `name` argument, similar to the events example, in which the `config.name` is only used if the `name` isn't passed.

File: templates/rust/http/src/handler.rs [hunk 5]
```diff
@@ -19,16 +20,24 @@ mod tests {
...
+            resp.body().as_ref().unwrap()
+        );
     }
 
     #[actix_rt::test]
     async fn post() {
         let req = test::TestRequest::post().to_http_request();
-        let resp = index(req).await;
+        let config = HandlerConfig::default();
+        let resp = index(req, web::Data::new(config.clone())).await;
```
Comment (jcrossley3) [inline templates/rust/http/src/handler.rs:36]: That `.clone()` call is making me question my wanting a static above. :)

**Review by jcrossley3 — COMMENTED** [review]
Cool idea using config to un-hardcode the "world" string. Just a few questions...

File: templates/rust/http/src/handler.rs (hunk 4 above)
Comment (dejanb) [reply templates/rust/http/src/handler.rs:11]: I'm all for improving this example and looked into it. However, I didn't find an easy/simple way to add it. We'd need to use extractors for both query and body (depending whether get or post is used) and possibly another struct. Maybe we should split the two handlers? We would also need to add more tests.
I would definitely evolve this, but maybe we should do it in a separate PR?

File: templates/rust/events/src/handler.rs (hunk 2 above)
Comment (dejanb) [reply templates/rust/events/src/handler.rs]: I ended up adding `config()` function to tests that wraps this. This is the simplest thing I could find as making static traits complicates them for the example and I would like to keep those as simple and reusable as possible. Let me know how it looks.

File: templates/rust/http/src/handler.rs (hunk 4 above)
Comment (jcrossley3) [reply templates/rust/http/src/handler.rs:11]: Yeah, I was thinking there was a `FromRequest` impl for `String` but that returns the whole body. I'm not sure introducing a `Query` param (or whatever) is worth it. Probably out of scope for an example. My bad!

File: templates/rust/events/src/handler.rs (hunk 2 above)
Comment (jcrossley3) [reply templates/rust/events/src/handler.rs]: Much better, thanks!

**Comment by lance** [comment]
Thanks for helping out with the review @jcrossley3 

/lgtm

**Comment by lance** [comment]
@dejanb pkged.go is such a pita. Could you please rebase this PR?

**Comment by dejanb** [comment]
> @dejanb pkged.go is such a pita. Could you please rebase this PR?

@lance sure, just did it. Tests look good now. Let me know if there's anything else needed.

**Pull request merged**